File: November/7/isomorphicstring.py
# LeetCode 205: Isomorphic Strings
# Explanation:
# 1. Maintain two maps to store character mappings from s→t and t→s.
# 2. Ensure one-to-one correspondence between both.
# 3. If mismatch found, return False.
# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        map_s_t, map_t_s = {}, {}
        for c1, c2 in zip(s, t):
            if (c1 in map_s_t and map_s_t[c1] != c2) or (c2 in map_t_s and map_t_s[c2] != c1):
                return False
            map_s_t[c1] = c2
            map_t_s[c2] = c1
        return True

    def isIsomorphicOneLine(self, s: str, t: str) -> bool:
        return [s.index(c) for c in s] == [t.index(c) for c in t]
    
sol = Solution()
print(sol.isIsomorphic("egg", "add"))    # True
print(sol.isIsomorphic("foo", "bar"))    # False
print(sol.isIsomorphic("paper", "title"))# True

File: 2025/November/7/isomorphicstring.cpp
// LeetCode 205: Isomorphic Strings
// Explanation:
// 1. Use two arrays to store mapping from s→t and t→s.
// 2. Validate bijection by ensuring consistent mapping both ways.
// Time Complexity: O(n)
// Space Complexity: O(1) (constant 256 chars)

#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) return false;
        vector<int> mapS(256, -1), mapT(256, -1);
        for (int i = 0; i < s.size(); ++i) {
            if (mapS[s[i]] == -1 && mapT[t[i]] == -1) {
                mapS[s[i]] = t[i];
                mapT[t[i]] = s[i];
            } else if (mapS[s[i]] != t[i] || mapT[t[i]] != s[i]) {
                return false;
            }
        }
        return true;
    }
};

/*
Solution sol;
cout << boolalpha;
cout << sol.isIsomorphic("egg", "add") << endl;    // true
cout << sol.isIsomorphic("foo", "bar") << endl;    // false
cout << sol.isIsomorphic("paper", "title") << endl;// true
*/